docs/developer-docs/security/formal-verification.md

```diff
@@ -0,0 +1,26 @@
+
+There are many existing formal verification tools. While none of them take into account the specifics of the Internet Computer yet, many of them are general enough that they can be applied to canisters. In particular, the DFINITY team has made good use of the TLA+ toolkit to eliminate reentrancy bugs, a type of concurrency bug where one method of a particular canister is called while another one is still executing. These bugs are particularly difficult to find, as they can involve unexpected interactions of code scattered throughout a canister, or sometimes even in different canisters altogether. The number of such code interactions may be huge, and thus difficult for humans to detect and weed out the unwanted ones. Moreover, these interactions are usually difficult to test automatically and systematically, which TLA+ can do.
+
+## TLA+
+
+The Temporal Logic of Actions (TLA+) is a language for specifying and verifying complex systems. TLA+ comes with a set of tools for lightweight formal verification in the form of so-called model checking. Through model checking, it exhaustively explores all possible concurrent interactions of a model of the code — exactly the domain that is difficult to test — and finds bugs. Importantly, after building the model of the code, model checking runs with virtually no further human input, making it highly cost-effective. To illustrate with some made-up numbers: if the industry standard practices (such as testing and security reviews) eliminate 80% of the bugs, and “heavyweight” formal verification eliminates 99.99%, with TLA+ you can eliminate 90% with a fraction of the effort of the heavyweight verification.
+
+We have used TLA+ to create the following models that can be interesting for dapp developers:
+
+- NNS and SNS governance (focusing on interactions with the ledger canister)
+- ICP ledger (focusing on block archival)
+- ckBTC minter
+- SNS swap canister
+- People parties dapp
+
+You can find more details on the benefits, high-level approach, and the low-level details of applying TLA+ to canisters and dapps in a series of blog posts ([1](https://medium.com/dfinity/eliminating-smart-contract-bugs-with-tla-e986aeb6da24), [2](https://medium.com/dfinity/weeding-out-the-bugs-with-tla-models-3606045bf24e), [3](https://mynosefroze.com/blog/2023-08-09-tla_for_canisters)), and the models themselves are [open source](https://github.com/dfinity/formal-models).
```

Comment:
Maybe we can add it something like " to find out more info that can also help you to use models for your canisters" (or so). But this is just an idea.

Reply:
Tried to do something now, I hope I got the direction you were aiming for right.